cordapp-cpk/src/main/kotlin/net/corda/plugins/cpb/CpbPlugin.kt

```diff
@@ -0,0 +1,120 @@
+        private const val CPB_TASK_NAME = "cpb"
+        private const val CPB_CONFIGURATION_NAME = CPB_TASK_NAME
+        private const val CPB_PACKAGING_CONFIGURATION_NAME = "cpbPackaging"
+        private const val CPB_ARTIFACT_CLASSIFIER = "package"
+        private const val CPB_FILE_EXTENSION = CPB_TASK_NAME
+        private const val CPK_FILE_SUFFIX = ".$CPK_FILE_EXTENSION"
+        private const val JAR_FILE_SUFFIX = ".jar"
+        private val EXCLUDED_CPK_TYPES = setOf("corda-api")
+    }
+
+    override fun apply(project: Project) {
+        project.pluginManager.apply(CordappPlugin::class.java)
+        val allCordappsConfiguration = project.configurations.getByName(ALL_CORDAPPS_CONFIGURATION_NAME)
+
+        val cpbConfiguration = project.configurations.create(CPB_CONFIGURATION_NAME)
+            .extendsFrom(allCordappsConfiguration)
+            .setVisible(false)
+            .apply {
+                isCanBeConsumed = false
+                isCanBeResolved = false
+            }
+
+        val cpbPackaging = project.configurations.create(CPB_PACKAGING_CONFIGURATION_NAME)
+            .setTransitive(false)
+            .setVisible(false)
+            .withDependencies { dependencies ->
+                // Force Gradle to discover any transitive CPKs
+                // and include them in cpbConfiguration.
+                project.configurations.detachedConfiguration()
+                    .setTransitive(false)
+                    .setVisible(false)
+                    .extendsFrom(cpbConfiguration)
+                    .resolvedConfiguration
+
+                val dependencyHandler = project.dependencies
+                cpbConfiguration.allDependencies.filterIsInstance(ModuleDependency::class.java).forEach { dependency ->
+                    when (dependency) {
+                        is ExternalDependency -> {
+                            val cpk = dependency.copy()
+                            cpk.artifact {
+                                it.name = dependency.name
+                                it.classifier = CPK_ARTIFACT_CLASSIFIER
+                                it.type = CPK_FILE_EXTENSION
+                            }
+                            dependencies.add(cpk)
+                        }
+                        is ProjectDependency -> {
+                            val cpk = dependencyHandler.project(mapOf<String, Any?>(
+                                "path" to dependency.dependencyProject.path,
+                                "configuration" to CORDA_CPK_CONFIGURATION_NAME
+                            ))
+                            dependencies.add(cpk)
+                        }
+                    }
+                }
+            }.apply {
+                isCanBeConsumed = false
+            }
+
+        project.afterEvaluate {
+            // We MUST resolve the CPB configurations before
+            // Gradle builds the Task Execution Graph!
+            cpbPackaging.resolve()
+        }
```

Comment:
why do we need to do this?

Reply:
Because of [Gradle #17765](https://github.com/gradle/gradle/issues/17765) :scream:

Reply:
BTW, are you going to rebase this branch? I have an extra tweak to apply now that #446 has been merged.

Reply:
done

Reply:
With any luck, we will only need this `afterEvaluate` handler for some versions of Gradle < 7.x (for unknown value of x...).